### ISSUE TITLE
The mAP for res3d50 model is much lower than 0.6

### ISSUE THREAD

**Issue by wwnbbd** [zhoubolei/moments_models#30]
I tried to use "test_video.py" file to load the pretrained "multi_moments_resnet3d50_wlsep.pth.tar" parameters and test on the validation set of multi-moments in time dataset. However, the mAP result showed by torchnet.metric is 0.3948 which is noticeable lower than 0.6 reported in the paper. 

In the test script, I set the frames to 16 and use [decord library](https://github.com/dmlc/decord) to extract frames from video.  Before feeding the data to model, basic preprocess is done, like resize and normalize. And after calculating, I use torchnet's mAP and AP meter to do mAP calculating. However, the result is frustrating, I got 0.4 instead of roughly 0.6. 

So, is there anything I'm missing?

```Python
"""Test pre-trained RGB model on a single video.

Date: 01/15/18
Authors: Bolei Zhou and Alex Andonian

This script accepts an mp4 video as the command line argument --video_file
and averages ResNet50 (trained on Moments) predictions on num_segment equally
spaced frames (extracted using ffmpeg).

Alternatively, one may instead provide the path to a directory containing
video frames saved as jpgs, which are sorted and forwarded through the model.

ResNet50 trained on Moments is used to predict the action for each frame,
and these class probabilities are average to produce a video-level predction.

Optionally, one can generate a new video --rendered_output from the frames
used to make the prediction with the predicted category in the top-left corner.

"""

import os
import argparse
#import moviepy.editor as mpy

import torch.optim
import torch.nn.parallel
from torch.nn import functional as F

import models
from utils import extract_frames, load_frames, render_frames
from torch.utils.data import DataLoader,Dataset

from decord import VideoReader
from decord import cpu, gpu
import numpy as np
import PIL
from PIL import Image
import torch
import torchnet.meter as meter


# options
parser = argparse.ArgumentParser(description="test TRN on a single video")
group = parser.add_mutually_exclusive_group(required=True)
group.add_argument('--video_file', type=str, default=None)
group.add_argument('--frame_folder', type=str, default=None)
parser.add_argument('--rendered_output', type=str, default=None)
parser.add_argument('--num_segments', type=int, default=16)
parser.add_argument('--arch', type=str, default='resnet3d50', choices=['resnet50', 'resnet3d50'])
args = parser.parse_args()

# Load model
model = models.load_model(args.arch)

# Get dataset categories
categories = models.load_categories()

# Load the video frame transform
transform = models.load_transform()

# Obtain video frames
"""
if args.frame_folder is not None:
    print('Loading frames in {}'.format(args.frame_folder))
    import glob
    # here make sure after sorting the frame paths have the correct temporal order
    frame_paths = sorted(glob.glob(os.path.join(args.frame_folder, '*.jpg')))
    frames = load_frames(frame_paths)
    print("Loading videos in {}".format(args.frame_folder))
else:
    print('Extracting frames using ffmpeg...')
    frames = extract_frames(args.video_file, args.num_segments)
"""
data_path = "/workdir/wwn/Multi_Moments_in_Time/videos/"
annotation_path = "./validationSet.txt"
class TestVideo(Dataset):
    def __init__(self, data_path, annotation_path):
        self.data_path = data_path
        self.records = []
        with open(annotation_path) as f:
            lines = f.readlines()
            for line in lines:
                self.records.append(line.strip().replace("\n",""))

    def __len__(self):
        return len(self.records)

    def __getitem__(self, index):
        single_record = self.records[index]
        parts = single_record.split(",")
        video_path = parts[0]
        video_path = os.path.join(self.data_path, video_path)
        vr = VideoReader(video_path, ctx=cpu(0))
        gap = int(len(vr)/16) #TODO change this to 5(5 fps) to see results
        index_list = [i * gap for i in range(16)]
        frames = vr.get_batch(index_list).asnumpy()
        frames = [Image.fromarray(frames[i]).convert("RGB") for i in range(16)]
        frames = torch.stack([transform(frame) for frame in frames],1)

        label = [0] * 313
        target_classes = parts[1:]
        for item in target_classes:
            label[int(item)] = 1
        label = torch.LongTensor(label)
        return {"frames":frames, "label":label}

"""
# Prepare input tensor
if args.arch == 'resnet3d50':
    # [1, num_frames, 3, 224, 224]
    input = torch.stack([transform(frame) for frame in frames], 1).unsqueeze(0)
else:
    # [num_frames, 3, 224, 224]
    input = torch.stack([transform(frame) for frame in frames])
"""
#dataset
val_dataset = TestVideo(data_path, annotation_path)
val_loader = DataLoader(val_dataset, batch_size=1, shuffle=False, num_workers=4, pin_memory=True)

device = torch.device("cuda", 0)
model = model.to(device)

mAP = meter.mAPMeter()
mAP.reset()
AP = meter.APMeter()
AP.reset()

# Make video prediction
for index, data in enumerate(val_loader):
    frames = data["frames"].to(device)
    target = data["label"].to(device)
    with torch.no_grad():
        logits = model(frames)
        
        #h_x = F.softmax(logits, 1).mean(dim=0)#TODO check output value range
        #probs, idx = h_x.sort(0, True)
        mAP.add(logits.data, target.data)
        AP.add(logits.data, target.data)
print(mAP.value())
print(AP.value())
np.save("res3d50-ap.npy", AP.value().data.cpu().numpy())

"""
# Output the prediction.
video_name = args.frame_folder if args.frame_folder is not None else args.video_file
print('RESULT ON ' + video_name)
for i in range(0, 5):
    print('{:.3f} -> {}'.format(probs[i], categories[idx[i]]))

# Render output frames with prediction text.
if args.rendered_output is not None:
    prediction = categories[idx[0]]
    rendered_frames = render_frames(frames, prediction)
    clip = mpy.ImageSequenceClip(rendered_frames, fps=4)
    clip.write_videofile(args.rendered_output)
"""
```
